File: clients/typescript/test/client.test.ts
// SPDX-FileCopyrightText: 2026 Olivares.AI
// SPDX-License-Identifier: Apache-2.0

// Smoke tests for the TypeScript client against a local fake control plane
// (node:http). Run from clients/typescript: pnpm test
// (`task sdk:test:ts` from the repo root.)

import { createServer, type IncomingMessage, type Server } from "node:http";
import { afterAll, beforeAll, beforeEach, describe, expect, it } from "vitest";
import {
  APIError,
  type AuthCapabilityQuestions,
  type AuthCapabilityResult,
  type AuthCapabilityResults,
  Client,
  type DeprecationNotice,
  type JsonInput,
  type SessionsCommunicationMessageSendBody,
} from "../src/index.js";

interface Seen {
  method: string;
  url: string;
  headers: IncomingMessage["headers"];
}

let server: Server;
let endpoint: string;
const seen: Seen[] = [];
let rateCalls = 0;

// A faithful answer from GET /v1/m/finops/statements/{id}/export: the handler's nine
// columns and one line. 9007199254740993 is 2**53+1 — the oracle, not filler. A
// consumer that routed this value through a JSON number and an IEEE double would hand
// back 9007199254740992.
const STATEMENT_EXPORT_CSV =
  "cost_center_code,cost_center_name,model,provider,agent," +
  "input_tokens,output_tokens,cost_micro_usd,sample_count\n" +
  "ENG-01,Engineering,claude-opus-5,anthropic,,100,50,9007199254740993,1\n";

beforeAll(async () => {
  server = createServer((req, res) => {
    seen.push({
      method: req.method ?? "",
      url: req.url ?? "",
      headers: req.headers,
    });
    const url = new URL(req.url ?? "/", "http://x");
    const json = (
      status: number,
      payload: unknown,
      headers: Record<string, string | string[]> = {},
    ) => {
      res.writeHead(status, { "Content-Type": "application/json", ...headers });
      res.end(JSON.stringify(payload));
    };
    if (url.pathname === "/v1/agents" && req.method === "GET") {
      if (url.searchParams.get("cursor") === "c1") {
        json(200, { items: [{ id: "c" }], has_more: false });
      } else {
        json(200, {
          items: [{ id: "a" }, { id: "b" }],
          cursor: "c1",
          has_more: true,
        });
      }
    } else if (url.pathname.startsWith("/v1/agents/")) {
      json(
        404,
        { error: { code: "not_found", message: "no such agent" } },
        { "X-Request-ID": "req-42" },
      );
    } else if (url.pathname === "/v1/server-info") {
      json(
        200,
        { version: "test" },
        {
          Deprecation: "@1780272000",
          Sunset: "Thu, 01 Jun 2028 00:00:00 GMT",
          Link: '<https://docs.olivares.invalid/how-to/migrate-example/>; rel="deprecation"',
        },
      );
    } else if (url.pathname === "/v1/tokens" && req.method === "POST") {
      rateCalls++;
      if (rateCalls === 1) {
        json(
          429,
          { error: { code: "rate_limited", message: "slow down" } },
          { "Retry-After": "3" },
        );
      } else {
        json(201, { ok: true });
      }
    } else if (
      url.pathname.startsWith("/v1/tokens/") &&
      req.method === "DELETE"
    ) {
      // Deprecated parameterized route: one notice per ENDPOINT expected.
      json(200, {}, { Deprecation: "@1780272000" });
    } else if (url.pathname === "/metrics") {
      res.writeHead(200, {
        "Content-Type": "text/plain; version=0.0.4; charset=utf-8",
      });
      res.end(
        "# HELP olivares_requests_total Requests.\nolivares_requests_total 42\n",
      );
    } else if (
      url.pathname.startsWith("/v1/m/finops/statements/") &&
      url.pathname.endsWith("/export")
    ) {
      // The chargeback statement export: always CSV on 200, never negotiated.
      res.writeHead(200, {
        "Content-Type": "text/csv; charset=utf-8",
        "Content-Disposition":
          'attachment; filename="chargeback_ENG-01_2026-06-01.csv"',
      });
      res.end(STATEMENT_EXPORT_CSV);
    } else if (url.pathname === "/raw-request") {
      json(200, {});
    } else if (url.pathname === "/v1/users" && req.method === "GET") {
      // Hostile shape: a 200 whose body is not a JSON object.
      json(200, [1, 2]);
    } else {
      json(400, { error: { code: "bad_request", message: "nope" } });
    }
  });
  await new Promise<void>((r) => server.listen(0, "127.0.0.1", r));
  const addr = server.address();
  if (addr === null || typeof addr === "string") throw new Error("no port");
  endpoint = `http://127.0.0.1:${addr.port}`;
});

afterAll(() => new Promise<void>((r) => server.close(() => r())));

beforeEach(() => {
  seen.length = 0;
  rateCalls = 0;
});

function newClient(
  extra: Partial<ConstructorParameters<typeof Client>[0]> = {},
) {
  return new Client({
    endpoint,
    token: "olvk_test_secret",
    tenant: "t-default",
    retrySleep: async (ms) => {
      slept.push(ms);
    },
    ...extra,
  });
}

class RawRequestClient extends Client {
  sendRaw(body: Uint8Array | undefined, contentType: string) {
    return this.doReqRawWithType(
      "POST",
      "/raw-request",
      "/raw-request",
      body,
      contentType,
    );
  }

  sendLegacyOctetStream(body: Uint8Array) {
    return this.doReqRaw("PUT", "/raw-request", "/raw-request", body);
  }
}

class JsonRequestClient extends Client {
  sendRequired(body: JsonInput) {
    return this.doJsonRequired("POST", "/required", "/required", body);
  }

  sendOptional(body?: JsonInput) {
    return this.do("POST", "/optional", "/optional", body);
  }
}
const slept: number[] = [];
beforeEach(() => {
  slept.length = 0;
});

describe("request shape", () => {
  it("sends auth, tenant override, UA and query", async () => {
    const c = newClient();
    const out = await c.getV1Agents({
      query: { limit: "5" },
      tenant: "t-override",
    });
    expect((out.items as { id: string }[]).map((i) => i.id)).toEqual([
      "a",
      "b",
    ]);
    const r = seen[0];
    expect(r.url).toBe("/v1/agents?limit=5");
    expect(r.headers.authorization).toBe("Bearer olvk_test_secret");
    expect(r.headers["x-olivares-tenant"]).toBe("t-override");
    expect(r.headers["user-agent"]).toContain("olivares-client-ts/");
    expect(r.headers["user-agent"]).toContain("(api v1)");
  });

  // The API has repeatable query parameters (GET /v1/audit's exclude_action, the
  // first of them). An array must become N occurrences: handed straight to
  // URLSearchParams it becomes ONE comma-joined value, which a server reading a
  // repeatable filter takes as a single value and matches nothing.
  it("repeats an array query param instead of comma-joining it", async () => {
    const c = newClient();
    await c.getV1Agents({ query: { limit: "5", tag: ["a", "b"] } });
    expect(seen[0].url).toBe("/v1/agents?limit=5&tag=a&tag=b");
  });

  it("escapes path params in generated operations", async () => {
    const c = newClient();
    await expect(c.getV1AgentsById("a/b c")).rejects.toThrow(APIError);
    expect(seen[0].url).toBe("/v1/agents/a%2Fb%20c");
  });

  it("preserves the declared raw request Content-Type", async () => {
    const c = new RawRequestClient({ endpoint, token: "olvk_test_secret" });
    await c.sendRaw(new TextEncoder().encode("{}\n"), "application/x-ndjson");
    expect(seen.at(-1)?.headers["content-type"]).toBe("application/x-ndjson");

    await c.sendLegacyOctetStream(new TextEncoder().encode("raw"));
    expect(seen.at(-1)?.headers["content-type"]).toBe(
      "application/octet-stream",
    );

    await c.sendRaw(undefined, "application/x-ndjson");
    expect(seen.at(-1)?.headers["content-type"]).toBeUndefined();
  });

  it("sends required JSON null but omits optional undefined", async () => {
    const requests: RequestInit[] = [];
    const c = new JsonRequestClient({
      endpoint: "https://olivares.invalid",
      fetch: async (_input, init) => {
        requests.push(init ?? {});
        return new Response("{}", {
          status: 200,
          headers: { "Content-Type": "application/json" },
        });
      },
    });

    const validInputs: JsonInput[] = [null, true, 1, "scalar", [null]];
    expect(validInputs).toHaveLength(5);
    await c.sendRequired(null);
    await c.sendOptional(undefined);

    expect(requests[0].body).toBe("null");
    expect(
      (requests[0].headers as Record<string, string>)["Content-Type"],
    ).toBe("application/json");
    expect(requests[1].body).toBeUndefined();
    expect(
      (requests[1].headers as Record<string, string>)["Content-Type"],
    ).toBeUndefined();
  });
});

describe("error envelope", () => {
  it("maps the single envelope to APIError", async () => {
    const c = newClient();
    const err = await c.getV1AgentsById("missing").catch((e: unknown) => e);
    expect(err).toBeInstanceOf(APIError);
    const ae = err as APIError;
    expect([ae.status, ae.code, ae.requestId]).toEqual([
      404,
      "not_found",
      "req-42",
    ]);
    expect(ae.message).toContain("no such agent");
  });
});

describe("retry policy", () => {
  it("retries 429 honouring Retry-After as the lower bound", async () => {
    const c = newClient();
    const out = await c.postV1Tokens({ name: "ci" });
    expect(out.ok).toBe(true);
    expect(rateCalls).toBe(2);
    expect(slept).toEqual([3000]);
  });

  it("never retries a 400", async () => {
    const c = newClient();
    await expect(c.postV1Memberships({})).rejects.toThrow(APIError);
    expect(seen.length).toBe(1);
  });
});

describe("stability policy signal", () => {
  it("surfaces deprecation headers once per endpoint", async () => {
    const notices: DeprecationNotice[] = [];
    const c = newClient({ onDeprecation: (n) => notices.push(n) });
    await c.getV1ServerInfo();
    await c.getV1ServerInfo();
    expect(notices).toEqual([
      {
        method: "GET",
        path: "/v1/server-info",
        deprecation: "@1780272000",
        sunset: "Thu, 01 Jun 2028 00:00:00 GMT",
        link: "https://docs.olivares.invalid/how-to/migrate-example/",
      },
    ]);
  });

  it("dedups per route template, not per resource", async () => {
    const notices: DeprecationNotice[] = [];
    const c = newClient({ onDeprecation: (n) => notices.push(n) });
    await c.deleteV1TokensById("tok_001");
    await c.deleteV1TokensById("tok_002");
    expect(notices).toHaveLength(1);
    expect(notices[0].path).toBe("/v1/tokens/tok_001");
  });
});

describe("response shape", () => {
  it("returns the raw body for non-JSON operations", async () => {
    const c = newClient();
    const text = await c.getMetrics();
    expect(text).toContain("olivares_requests_total 42");
    expect(seen[0].headers.accept).not.toBe("application/json");
  });

  it("consumes a CSV 200 through the generated statement export operation", async () => {
    // ⛔ While the document declared this 200 an application/json object the emitter
    // produced `Promise<Json>` + this.do, and core.ts ran JSON.parse over the CSV: a
    // valid 200 raised APIError(bad_response) in every generated TypeScript client.
    // The assertion is on the OPERATION — doRaw itself was always right.
    const c = newClient();
    const text = await c.getV1MFinopsStatementsByIdExport(
      "01a084d5-988c-7e46-b396-5cff04bf2793",
    );
    expect(text).toBe(STATEMENT_EXPORT_CSV);
    expect(text).toContain("9007199254740993");
    expect(seen[0].url).toBe(
      "/v1/m/finops/statements/01a084d5-988c-7e46-b396-5cff04bf2793/export",
    );
    // A CSV-only route must not demand JSON: the server does not negotiate.
    expect(seen[0].headers.accept).not.toBe("application/json");
  });

  it("rejects a 200 whose body is not a JSON object", async () => {
    const c = newClient();
    const err = await c.getV1Users().catch((e: unknown) => e);
    expect(err).toBeInstanceOf(APIError);
    expect((err as APIError).code).toBe("bad_response");
  });
});

describe("pagination", () => {
  it("follows the items/cursor/has_more envelope", async () => {
    const c = newClient();
    const ids: string[] = [];
    for await (const item of c.paginate("/v1/agents")) {
      ids.push(item.id as string);
    }
    expect(ids).toEqual(["a", "b", "c"]);
  });
});

describe("construction", () => {
  it("publishes the typed communication request surface from the package root", () => {
    const body: SessionsCommunicationMessageSendBody = {
      channel_id: "00000000-0000-4000-8000-000000000001",
      recipient: { kind: "user", ref: "00000000-0000-4000-8000-000000000002" },
      content: {
        subject: "scheduled",
        blocks: [{ type: "text", text: "hello" }],
      },
      available_at: "2026-09-06T03:00:00Z",
    };
    expect(body.available_at).toBe("2026-09-06T03:00:00Z");
  });

  it("rejects a non-absolute endpoint", () => {
    expect(() => new Client({ endpoint: "not-a-url" })).toThrow();
  });
});

describe("capability projection, schema 2", () => {
  // docs/contracts/CAPABILITY-PROJECTION.md: one positive with its budget, one
  // concealed non-verdict without one, one surface admission.
  const published: AuthCapabilityResults = {
    schema_version: 2,
    results: [
      {
        id: "sheet",
        kind: "operation",
        state: "allowed",
        code: "authorized",
        observed_at: "2026-09-07T10:00:00.250Z",
        refresh_after_ms: 30000,
      },
      {
        id: "held",
        kind: "operation",
        state: "undisclosed",
        code: "not_disclosed",
        observed_at: "2026-09-07T10:00:00Z",
      },
      {
        id: "list",
        kind: "surface",
        state: "reachable",
        code: "admitted",
        observed_at: "2026-09-07T10:00:00.100Z",
        refresh_after_ms: 12000,
      },
    ],
  };

  it("round-trips the generated DTO family through the package root", async () => {
    const requests: { url: string; init: RequestInit }[] = [];
    const c = new Client({
      endpoint: "https://olivares.invalid",
      token: "olvk_test_secret",
      fetch: async (input, init) => {
        requests.push({ url: String(input), init: init ?? {} });
        return new Response(JSON.stringify(published), {
          status: 200,
          headers: {
            "Content-Type": "application/json",
            "Cache-Control": "no-store",
          },
        });
      },
    });
    const workspace = "00000000-0000-4000-8000-000000000010";
    const body: AuthCapabilityQuestions = {
      schema_version: 2,
      questions: [
        {
          id: "sheet",
          kind: "operation",
          operation: "GET /v1/m/sessions/channels/{id}/grants",
          workspace_id: workspace,
          selectors: { path: { id: "00000000-0000-4000-8000-000000000001" } },
        },
        {
          id: "held",
          kind: "operation",
          operation: "PATCH /v1/m/sessions/channels",
          workspace_id: workspace,
          selectors: {
            body: { channel_id: "00000000-0000-4000-8000-000000000002" },
          },
        },
        {
          id: "list",
          kind: "surface",
          operation: "GET /v1/m/sessions/channels",
          workspace_id: workspace,
        },
      ],
    };
    const out: AuthCapabilityResults = await c.postV1AuthCapabilities({ body });

    expect(requests).toHaveLength(1);
    expect(requests[0].url).toBe(
      "https://olivares.invalid/v1/auth/capabilities",
    );
    expect(requests[0].init.method).toBe("POST");
    expect(
      (requests[0].init.headers as Record<string, string>)["Content-Type"],
    ).toBe("application/json");
    const wire = JSON.parse(String(requests[0].init.body)) as Record<
      string,
      unknown
    >;
    expect(wire.schema_version).toBe(2);
    expect(wire.questions).toEqual(body.questions);

    expect(out.schema_version).toBe(2);
    expect(out.results.map((r) => [r.id, r.state, r.code])).toEqual([
      ["sheet", "allowed", "authorized"],
      ["held", "undisclosed", "not_disclosed"],
      ["list", "reachable", "admitted"],
    ]);
    const positive: AuthCapabilityResult = out.results[0];
    const concealed: AuthCapabilityResult = out.results[1];
    expect(positive.refresh_after_ms).toBe(30000);
    // The non-verdict carries no budget at all: absent, not zero.
    expect("refresh_after_ms" in concealed).toBe(false);
    expect(concealed.observed_at).toBe("2026-09-07T10:00:00Z");
    // Re-serializing the typed result reproduces the wire exactly.
    expect(JSON.parse(JSON.stringify(out))).toEqual(published);
  });
});

File: clients/go/client_test.go
// SPDX-FileCopyrightText: 2026 Olivares.AI
// SPDX-License-Identifier: Apache-2.0

package olivares

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"sync/atomic"
	"testing"
	"time"
)

// newTestClient builds a Client against a fake server with an instant,
// recording sleep (so retry waits are asserted, not slept).
func newTestClient(t *testing.T, h http.Handler, opts ...Option) (*Client, *[]time.Duration) {
	t.Helper()
	srv := httptest.NewServer(h)
	t.Cleanup(srv.Close)
	c, err := New(srv.URL, "olvk_test_secret", opts...)
	if err != nil {
		t.Fatal(err)
	}
	var slept []time.Duration
	c.sleep = func(_ context.Context, d time.Duration) error {
		slept = append(slept, d)
		return nil
	}
	return c, &slept
}

func TestRequestShape(t *testing.T) {
	var got *http.Request
	c, _ := newTestClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		got = r.Clone(context.Background())
		_, _ = w.Write([]byte(`{"items":[],"has_more":false}`))
	}), WithTenant("t-default"))

	if _, err := c.GetV1Agents(context.Background(), Query("limit", "5"), Tenant("t-override")); err != nil {
		t.Fatal(err)
	}
	if got.URL.Path != "/v1/agents" || got.URL.Query().Get("limit") != "5" {
		t.Errorf("request = %s %s", got.URL.Path, got.URL.RawQuery)
	}
	if h := got.Header.Get("Authorization"); h != "Bearer olvk_test_secret" {
		t.Errorf("Authorization = %q", h)
	}
	if h := got.Header.Get("X-Olivares-Tenant"); h != "t-override" {
		t.Errorf("tenant override lost: %q", h)
	}
	if ua := got.Header.Get("User-Agent"); !strings.HasPrefix(ua, "olivares-client-go/"+Version) ||
		!strings.Contains(ua, "api "+APIVersion) {
		t.Errorf("User-Agent = %q", ua)
	}
}

func TestRawRequestPreservesDeclaredContentType(t *testing.T) {
	var got []string
	c, _ := newTestClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		got = append(got, r.Header.Get("Content-Type"))
		_, _ = w.Write([]byte(`{}`))
	}))

	ctx := context.Background()
	if _, err := c.doReqRawWithType(
		ctx, http.MethodPost, "/memory/import", "/memory/import",
		[]byte("{}\n"), "application/x-ndjson",
	); err != nil {
		t.Fatal(err)
	}
	if _, err := c.doReqRaw(
		ctx, http.MethodPut, "/files/raw", "/files/raw", []byte("raw"),
	); err != nil {
		t.Fatal(err)
	}
	if _, err := c.doReqRawWithType(
		ctx, http.MethodPost, "/memory/import", "/memory/import", nil,
		"application/x-ndjson",
	); err != nil {
		t.Fatal(err)
	}

	want := []string{"application/x-ndjson", "application/octet-stream", ""}
	if strings.Join(got, ",") != strings.Join(want, ",") {
		t.Fatalf("Content-Type headers = %q, want %q", got, want)
	}
}

func TestRequiredJSONNullIsPresentWhileOptionalNilIsAbsent(t *testing.T) {
	type request struct {
		body        string
		contentType string
	}
	var got []request
	c, _ := newTestClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		body, err := io.ReadAll(r.Body)
		if err != nil {
			t.Errorf("read request body: %v", err)
		}
		got = append(got, request{body: string(body), contentType: r.Header.Get("Content-Type")})
		_, _ = w.Write([]byte(`{}`))
	}))

	ctx := context.Background()
	if _, err := c.doJSONRequired(ctx, http.MethodPost, "/required", "/required", nil); err != nil {
		t.Fatal(err)
	}
	if _, err := c.do(ctx, http.MethodPost, "/optional", "/optional", nil); err != nil {
		t.Fatal(err)
	}

	want := []request{
		{body: "null", contentType: "application/json"},
		{body: "", contentType: ""},
	}
	if len(got) != len(want) {
		t.Fatalf("requests = %#v, want %#v", got, want)
	}
	for i := range want {
		if got[i] != want[i] {
			t.Errorf("request %d = %#v, want %#v", i, got[i], want[i])
		}
	}
}

func TestPathEscapingThroughGeneratedOp(t *testing.T) {
	var gotPath string
	c, _ := newTestClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotPath = r.URL.EscapedPath()
		_, _ = w.Write([]byte(`{}`))
	}))
	if _, err := c.GetV1AgentsByID(context.Background(), "a/b c"); err != nil {
		t.Fatal(err)
	}
	if gotPath != "/v1/agents/a%2Fb%20c" {
		t.Errorf("escaped path = %q", gotPath)
	}
}

func TestErrorEnvelope(t *testing.T) {
	c, _ := newTestClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("X-Request-ID", "req-42")
		w.WriteHeader(http.StatusNotFound)
		_, _ = w.Write([]byte(`{"error":{"code":"not_found","message":"no such agent"}}`))
	}))
	_, err := c.GetV1AgentsByID(context.Background(), "missing")
	var ae *APIError
	if !errors.As(err, &ae) {
		t.Fatalf("err = %v, want *APIError", err)
	}
	if ae.Status != 404 || ae.Code != "not_found" || ae.Message != "no such agent" || ae.RequestID != "req-42" {
		t.Errorf("APIError = %+v", ae)
	}
}

func TestRetry429HonoursRetryAfter(t *testing.T) {
	var calls atomic.Int32
	c, slept := newTestClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if calls.Add(1) == 1 {
			w.Header().Set("Retry-After", "3")
			w.WriteHeader(http.StatusTooManyRequests)
			_, _ = w.Write([]byte(`{"error":{"code":"rate_limited","message":"slow down"}}`))
			return
		}
		_, _ = w.Write([]byte(`{"ok":true}`))
	}))
	out, err := c.PostV1Agents(context.Background(), map[string]any{"name": "a"})
	if err != nil || out["ok"] != true {
		t.Fatalf("out=%v err=%v", out, err)
	}
	if calls.Load() != 2 {
		t.Errorf("calls = %d, want 2 (one retry)", calls.Load())
	}
	if len(*slept) != 1 || (*slept)[0] != 3*time.Second {
		t.Errorf("slept = %v, want [3s] (Retry-After is the lower bound)", *slept)
	}
}

func TestRetryBudgetExhausts(t *testing.T) {
	var calls atomic.Int32
	c, _ := newTestClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		calls.Add(1)
		w.WriteHeader(http.StatusTooManyRequests)
		_, _ = w.Write([]byte(`{"error":{"code":"rate_limited","message":"still"}}`))
	}), WithMaxRetries(2))
	_, err := c.GetV1Agents(context.Background())
	var ae *APIError
	if !errors.As(err, &ae) || ae.Code != "rate_limited" {
		t.Fatalf("err = %v", err)
	}
	if calls.Load() != 3 {
		t.Errorf("calls = %d, want 3 (initial + 2 retries)", calls.Load())
	}
}

func Test503RetriesOnlyGET(t *testing.T) {
	var calls atomic.Int32
	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if calls.Add(1) == 1 {
			w.WriteHeader(http.StatusServiceUnavailable)
			_, _ = w.Write([]byte(`{"error":{"code":"not_leader","message":"handoff"}}`))
			return
		}
		_, _ = w.Write([]byte(`{}`))
	})

	c, _ := newTestClient(t, handler)
	if _, err := c.GetV1Agents(context.Background()); err != nil {
		t.Fatalf("GET should retry the 503 handoff: %v", err)
	}

	calls.Store(0)
	c2, _ := newTestClient(t, handler)
	if _, err := c2.PostV1Agents(context.Background(), map[string]any{}); err == nil {
		t.Fatal("POST must NOT retry a 503 (not known idempotent)")
	} else if calls.Load() != 1 {
		t.Errorf("POST calls = %d, want 1", calls.Load())
	}
}

func TestNoRetryOn400(t *testing.T) {
	var calls atomic.Int32
	c, _ := newTestClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		calls.Add(1)
		w.WriteHeader(http.StatusBadRequest)
		_, _ = w.Write([]byte(`{"error":{"code":"bad_request","message":"nope"}}`))
	}))
	if _, err := c.GetV1Agents(context.Background()); err == nil {
		t.Fatal("want error")
	}
	if calls.Load() != 1 {
		t.Errorf("calls = %d, want 1 (4xx is never retried)", calls.Load())
	}
}

func TestDeprecationNoticeOncePerEndpoint(t *testing.T) {
	var notices []DeprecationNotice
	c, _ := newTestClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Deprecation", "@1780272000")
		w.Header().Set("Sunset", "Thu, 01 Jun 2028 00:00:00 GMT")
		w.Header().Add("Link", `<https://docs.olivares.invalid/how-to/migrate-example/>; rel="deprecation"`)
		w.Header().Add("Link", `<https://docs.olivares.invalid/how-to/migrate-example/>; rel="sunset"`)
		_, _ = w.Write([]byte(`{}`))
	}), WithDeprecationHandler(func(n DeprecationNotice) { notices = append(notices, n) }))

	ctx := context.Background()
	for range 3 {
		if _, err := c.GetV1ServerInfo(ctx); err != nil {
			t.Fatal(err)
		}
	}
	if _, err := c.GetHealthz(ctx); err != nil {
		t.Fatal(err)
	}
	if len(notices) != 2 {
		t.Fatalf("notices = %d, want 2 (deduped per endpoint)", len(notices))
	}
	n := notices[0]
	if n.Method != "GET" || n.Path != "/v1/server-info" ||
		n.Deprecation != "@1780272000" || n.Sunset != "Thu, 01 Jun 2028 00:00:00 GMT" ||
		n.Link != "https://docs.olivares.invalid/how-to/migrate-example/" {
		t.Errorf("notice = %+v", n)
	}
}

func TestListPages(t *testing.T) {
	c, _ := newTestClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch r.URL.Query().Get("cursor") {
		case "":
			_ = json.NewEncoder(w).Encode(map[string]any{
				"items":  []any{map[string]any{"id": "a"}, map[string]any{"id": "b"}},
				"cursor": "c1", "has_more": true,
			})
		case "c1":
			_ = json.NewEncoder(w).Encode(map[string]any{
				"items": []any{map[string]any{"id": "c"}}, "has_more": false,
			})
		default:
			t.Errorf("unexpected cursor %q", r.URL.Query().Get("cursor"))
		}
	}))
	var ids []string
	for item, err := range c.ListPages(context.Background(), "/v1/agents") {
		if err != nil {
			t.Fatal(err)
		}
		ids = append(ids, item["id"].(string))
	}
	if strings.Join(ids, ",") != "a,b,c" {
		t.Errorf("ids = %v", ids)
	}
}

func TestDeprecationDedupPerRouteTemplate(t *testing.T) {
	var notices []DeprecationNotice
	c, _ := newTestClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Deprecation", "@1780272000")
		_, _ = w.Write([]byte(`{}`))
	}), WithDeprecationHandler(func(n DeprecationNotice) { notices = append(notices, n) }))

	ctx := context.Background()
	for _, id := range []string{"agt_001", "agt_002", "agt_003"} {
		if _, err := c.GetV1AgentsByID(ctx, id); err != nil {
			t.Fatal(err)
		}
	}
	if len(notices) != 1 {
		t.Fatalf("notices = %d, want 1 (deduped per route template, not per resource)", len(notices))
	}
	if notices[0].Path != "/v1/agents/agt_001" {
		t.Errorf("notice path = %q (the concrete path of the first hit)", notices[0].Path)
	}
}

func TestDeprecationLinkCoalescedHeader(t *testing.T) {
	var notices []DeprecationNotice
	c, _ := newTestClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Deprecation", "@1780272000")
		// A proxy may coalesce repeated Link headers into one comma-joined value.
		w.Header().Set("Link", `<https://cdn.example/x>; rel="preload", <https://docs.olivares.invalid/how-to/migrate-example/>; rel="deprecation"`)
		_, _ = w.Write([]byte(`{}`))
	}), WithDeprecationHandler(func(n DeprecationNotice) { notices = append(notices, n) }))
	if _, err := c.GetHealthz(context.Background()); err != nil {
		t.Fatal(err)
	}
	if len(notices) != 1 || notices[0].Link != "https://docs.olivares.invalid/how-to/migrate-example/" {
		t.Fatalf("notices = %+v, want the rel=deprecation target from the coalesced value", notices)
	}
}

func TestRawOperationMetrics(t *testing.T) {
	exposition := "# HELP olivares_requests_total Requests.\nolivares_requests_total 42\n"
	c, _ := newTestClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/metrics" {
			t.Errorf("path = %s", r.URL.Path)
		}
		if accept := r.Header.Get("Accept"); accept == "application/json" {
			t.Errorf("raw operation must not demand JSON (Accept=%q)", accept)
		}
		w.Header().Set("Content-Type", "text/plain; version=0.0.4; charset=utf-8")
		_, _ = w.Write([]byte(exposition))
	}))
	got, err := c.GetMetrics(context.Background())
	if err != nil {
		t.Fatal(err)
	}
	if string(got) != exposition {
		t.Errorf("GetMetrics = %q", got)
	}
}

func TestNewRejectsBadEndpoint(t *testing.T) {
	for _, bad := range []string{"", "not-a-url", "/relative"} {
		if _, err := New(bad, ""); err == nil {
			t.Errorf("New(%q) accepted a non-absolute endpoint", bad)
		}
	}
}

// statementExportCSV is a faithful answer from GET /v1/m/finops/statements/{id}/export:
// the handler's nine columns and one line. The 9007199254740993 is 2^53+1 — it is the
// oracle, not filler. Any consumer that routed this value through a JSON number and an
// IEEE double would hand back 9007199254740992.
const statementExportCSV = "cost_center_code,cost_center_name,model,provider,agent,input_tokens,output_tokens,cost_micro_usd,sample_count\n" +
	"ENG-01,Engineering,claude-opus-5,anthropic,,100,50,9007199254740993,1\n"

// TestStatementExportOperationConsumesCSV drives the GENERATED operation against a
// server that answers what the real handler answers.
//
// ⛔ THE DEFECT THIS CLOSES. While the document declared this 200 an application/json
// object, the generator emitted `map[string]any` + c.do, so this exact response died in
// json.Unmarshal with bad_response — a valid 200 that no generated Go client could read.
// The test is on the OPERATION, not on doRaw: the transport seam was always correct, and
// exercising it directly would have stayed green through the whole defect.
func TestStatementExportOperationConsumesCSV(t *testing.T) {
	var got *http.Request
	c, _ := newTestClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		got = r.Clone(context.Background())
		w.Header().Set("Content-Type", "text/csv; charset=utf-8")
		w.Header().Set("Content-Disposition", `attachment; filename="chargeback_ENG-01_2026-06-01.csv"`)
		_, _ = io.WriteString(w, statementExportCSV)
	}))

	body, err := c.GetV1MFinopsStatementsByIDExport(context.Background(), "01a084d5-988c-7e46-b396-5cff04bf2793")
	if err != nil {
		t.Fatalf("generated statement export operation failed on a valid CSV 200: %v", err)
	}
	if string(body) != statementExportCSV {
		t.Errorf("body = %q, want the CSV byte-for-byte", body)
	}
	if !strings.Contains(string(body), "9007199254740993") {
		t.Errorf("the exact micro-USD integer did not survive: %q", body)
	}
	if got.URL.Path != "/v1/m/finops/statements/01a084d5-988c-7e46-b396-5cff04bf2793/export" {
		t.Errorf("path = %q, want the id interpolated into the export route", got.URL.Path)
	}
	// A CSV-only route must not demand JSON: the server does not negotiate.
	if accept := got.Header.Get("Accept"); accept == "application/json" {
		t.Errorf("Accept = %q, want no JSON demand on a raw operation", accept)
	}
}
